agent/agent.go

```diff
@@ -0,0 +1,183 @@
+package agent
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/jfrog/jfrog-client-go/artifactory"
+	"github.com/jfrog/jfrog-client-go/artifactory/auth"
+	"github.com/jfrog/jfrog-client-go/artifactory/services"
+	aflog "github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/simplifi/looking-glass/config"
+	"log"
+	"os"
+	"path"
+	"time"
+)
+
+type Agent struct {
+	artifactoryManager artifactory.ArtifactoryServicesManager
+	awsSession session.Session
+	agentConfig config.AgentConfig
+	localStoragePath string
+}
+
+func New(artifactoryConfig config.ArtifactoryConfig, agentConfig config.AgentConfig) (*Agent, error) {
+	artMgr, err := createArtifactoryManager(artifactoryConfig.Url, artifactoryConfig.Key, artifactoryConfig.UserName)
+	if err != nil {
+		return nil, err
+	}
+	awsSess, err := createAwsSession(agentConfig.AwsKey, agentConfig.AwsSecret, agentConfig.AwsRegion)
+	if err != nil {
+		return nil, err
+	}
+	localStoragePath := path.Join("/tmp", agentConfig.Name)
+
+	agent := Agent{
+		artifactoryManager: *artMgr,
+		awsSession:         *awsSess,
+		agentConfig:        agentConfig,
+		localStoragePath:   localStoragePath,
+	}
+
+	return &agent, nil
+}
+
+func createArtifactoryManager(url string, apiKey string, userName string) (*artifactory.ArtifactoryServicesManager, error) {
+	// You have to setup a logger for Artifactory client to work
+	aflog.SetLogger(aflog.NewLogger(aflog.ERROR, nil))
+
+	details := auth.NewArtifactoryDetails()
+	details.SetUrl(url)
+	details.SetApiKey(apiKey)
+	details.SetUser(userName)
+
+	serviceConfig, err := artifactory.NewConfigBuilder().
+		SetArtDetails(details).
+		SetDryRun(false).
+		Build()
+	if err != nil {
+		return nil, err
+	}
+
+	mgr, err := artifactory.New(&details, serviceConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return mgr, nil
+}
+
+func createAwsSession(awsKey string, awsSecret string, awsRegion string) (*session.Session, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(awsRegion),
+		Credentials: credentials.NewStaticCredentials(awsKey, awsSecret, ""),
+	})
+
+	return sess, err
+}
+
+func (agt *Agent) Start() {
+	for {
+		for _, obj := range agt.getS3Objects() {
+
+			// see if any objects are missing from Artifactory
+			if !agt.existsInArtifactory(obj) {
+				log.Printf("INFO: [mirror] %s", obj)
+
+				// download object to local storage
+				localFile := path.Join(agt.localStoragePath, obj)
+				agt.downloadS3Obj(obj, localFile)
```

Comment:
Since this doesn't indicate success/failure, if it fails, the next step(s) will also fail, which I think would give us a bunch of misleading logs.